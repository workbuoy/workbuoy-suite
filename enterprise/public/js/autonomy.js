window.Autonomy=(function(){
  async function generateSlidesPDF(data){
    const tpl = await fetch('/templates/slides/3up.html').then(r=>r.text());
    const html = tpl
      .replaceAll('{{title}}', data.title||'WorkBuoy Slides')
      .replaceAll('{{slide1_title}}', data.slide1_title||'Slide 1')
      .replaceAll('{{slide2_title}}', data.slide2_title||'Slide 2')
      .replaceAll('{{slide3_title}}', data.slide3_title||'Slide 3')
      .replace('{{slide1_bullets}}', (data.slide1_bullets||[]).map(x=>`<li>${x}</li>`).join(''))
      .replace('{{slide2_bullets}}', (data.slide2_bullets||[]).map(x=>`<li>${x}</li>`).join(''))
      .replace('{{slide3_bullets}}', (data.slide3_bullets||[]).map(x=>`<li>${x}</li>`).join(''));
    const w=window.open('', '_blank'); w.document.write(html); w.document.close(); setTimeout(()=>w.print(), 250);
  }
  function renewalBrief(account){
    return [
      `Kunde: ${account.name} (${account.city})`,
      `Status: ${account.status}; fornyelse om ${account.renewal_days} dager`,
      `Mål: sikre fornyelse, avdekke oppgraderingsmulighet`,
      `Agenda: 1) Behov, 2) Verdi, 3) Neste steg`
    ].join('\n• ');
  }
  return {generateSlidesPDF,renewalBrief};
})();