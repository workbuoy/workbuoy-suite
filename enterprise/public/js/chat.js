(function(){
  const logEl=document.getElementById('chat-log'); const input=document.getElementById('chat-input'); const sendBtn=document.getElementById('send-btn'); const chipsEl=document.getElementById('chips');
  const modeSel=document.getElementById('coreMode'); let msgCount=0, kitShown=localStorage.getItem('wb.kitPitchShown')==='1';
  async function initCoreModes(){
    const modes = await fetch('/config/core.config.json').then(r=>r.json());
    modeSel.innerHTML = modes.map(m=>`<option value="${m.id}">${m.label}</option>`).join('');
    const saved = localStorage.getItem('wb.coreMode') || 'proactive'; modeSel.value=saved;
    modeSel.onchange = ()=>localStorage.setItem('wb.coreMode', modeSel.value);
  }
  function add(from,text){ const d=document.createElement('div'); d.className='msg'; d.innerHTML='<strong>'+from+':</strong> '+Secure.maskText(text); logEl.appendChild(d); logEl.scrollTop=logEl.scrollHeight; }
  function replyFor(text){
    const m = (localStorage.getItem('wb.coreMode')||'proactive');
    if(m==='invisible') return '(Usynlig) — ingen forslag.';
    if(m==='calm') return 'Jeg svarer når du spør — hva trenger du?';
    if(text.toLowerCase().includes('brief')) return 'Her er en kort fornyelses‑brief. Skal jeg lage agenda også?';
    if(text.toLowerCase().includes('agenda')) return 'Agenda kl. 14:00: 1) Behov 2) Verdi 3) Neste steg — vil du ha e‑postutkast etterpå?';
    return 'Notert. Vil du ha plan eller utkast?';
  }
  async function send(){
    const text=input.value.trim(); if(!text) return;
    MetaLearn.record({type:'chat_send', len:text.length});
    add('Du', text); input.value=''; msgCount++;
    const reply = replyFor(text);
    add('Buoy', reply);
    if(!kitShown && msgCount>=14){ kitShown=true; localStorage.setItem('wb.kitPitchShown','1'); Kits.showKitModal(); }
  }
  async function sendQuick(text){
    input.value = text; await send();
    if(text.toLowerCase().includes('fornyelses-brief')){
      const acct = window.__lastAccount || {name:'ACME Inc.', city:'Bergen', status:'Warm', renewal_days:30};
      add('Buoy', 'Ferdig — her er utkastet:\n• '+Autonomy.renewalBrief(acct));
    }
    if(text.toLowerCase().includes('agenda')){
      add('Buoy','Agenda (15 min)\n1) Status & mål\n2) Behov / Verdi\n3) Avtalt neste steg');
    }
  }
  window.Chat = {sendQuick};
  sendBtn.onclick=send;
  input.addEventListener('keydown',e=>{ if(e.key==='Enter' && !e.shiftKey){ e.preventDefault(); send(); }});
  Secure.init(); initCoreModes();
})();