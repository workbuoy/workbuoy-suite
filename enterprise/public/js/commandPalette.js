(function(){
  window.addEventListener('keydown', e=>{
    if((e.metaKey||e.ctrlKey) && e.shiftKey && e.key.toLowerCase()==='p'){
      e.preventDefault(); document.getElementById('chat-input').focus();
    }
  });
})();