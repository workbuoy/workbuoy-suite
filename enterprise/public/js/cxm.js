(function(){
  const tabs=['CRM','Analytics','Planner','News','Finance','Procurement','Projects'];
  const tabsEl=document.getElementById('cxm-tabs'); const body=document.getElementById('cxm-body');
  function btn(label){ const b=document.createElement('button'); b.className='chip'; b.textContent=label; return b; }
  tabs.forEach(t=>{ const b=btn(t); b.onclick=()=>select(t); tabsEl.appendChild(b); });

  async function select(t){
    if(t==='Finance'){ await renderModule('/api/cxm/finance', d=>{
      return `<div class="msg"><strong>MRR:</strong> ${d.mrr} — <strong>Burn:</strong> ${d.burn} — <strong>Cash:</strong> ${d.cash_on_hand}</div>`;
    }); }
    else if(t==='Procurement'){ await renderModule('/api/cxm/procurement', d=>{
      const rfqs = d.rfqs.map(r=>`<li>${r.id}: ${r.item} — ${r.status} (${r.bids} bud)</li>`).join('');
      const sl = d.shortlist.map(v=>`<li>${v.vendor}: ${v.score}</li>`).join('');
      return `<div class="msg"><strong>RFQs</strong><ul>${rfqs}</ul><strong>Shortlist</strong><ul>${sl}</ul></div>`;
    }); }
    else if(t==='Analytics'){ await renderModule('/api/cxm/analytics', d=>{
      const k = d.kpis.map(k=>`<div class="kpi">${k.name}: ${k.value}</div>`).join('');
      return `<div class="msg">${k}</div>`;
    }); }
    else if(t==='News'){ await renderModule('/api/cxm/news', d=>{
      const items = d.items.map(n=>`<li>${n.ts}: ${n.title} <em>(${n.source})</em></li>`).join('');
      return `<div class="msg"><strong>Company feed</strong><ul>${items}</ul></div>`;
    }); }
    else if(t==='CRM'){
      const r=await fetch('/assets/data/crm.json'); const data=await r.json();
      body.innerHTML = data.accounts.map(a=>`<div class="msg crm-row" data-id="${a.id}">${a.name} — ${a.industry} — ${a.status} — renewal: ${a.renewal_days}d</div>`).join('');
      body.querySelectorAll('.crm-row').forEach(row=>row.onclick=()=>{
        const id=row.getAttribute('data-id'); const acct=data.accounts.find(x=>x.id===id);
        window.__lastAccount = acct;
        const evt={type:'crm:renewal_due', ts:new Date().toISOString(), urgency:.9, impact:.8, title:'Forbered fornyelsessamtale', account:acct};
        EventBus.emit(evt);
        const chips=document.getElementById('chips'); chips.innerHTML='';
        ['Lag fornyelses-brief','Lag agenda for samtale','Utkast til oppfølging'].forEach(c=>{
          const el=document.createElement('div'); el.className='chip'; el.textContent=c; el.onclick=()=>window.Chat?.sendQuick(c); chips.appendChild(el);
        });
      });
    } else {
      body.innerHTML = '<div class="msg">Viser '+t+' (demo)</div>';
      if(t==='Planner'){ EventBus.emit({type:'planner:deadline_today', ts:new Date().toISOString(),urgency:.8,impact:.6,title:'Møte kl. 14:00—lag agenda'}); }
    }
  }

  async function renderModule(url, tpl){
    body.innerHTML = '<div class="msg">Henter data…</div>';
    const r = await fetch(url); const json = await r.json();
    body.innerHTML = tpl(json.data);
    // Show topp-level signal chip area (already exists)
  }

  select('CRM');
})();
