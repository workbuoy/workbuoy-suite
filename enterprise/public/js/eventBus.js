window.EventBus=(function(){
  const listeners=new Map(); const recent=[]; const MAX=400;
  function on(type,fn){ if(!listeners.has(type)) listeners.set(type,[]); listeners.get(type).push(fn); }
  function emit(evt){ recent.push(evt); if(recent.length>MAX) recent.shift(); (listeners.get(evt.type)||[]).forEach(fn=>fn(evt)); }
  function last(n=50){ return recent.slice(-n); }
  return {on,emit,last};
})();