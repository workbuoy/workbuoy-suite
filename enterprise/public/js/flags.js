window.FLAGS=(function(){
  const o={ focus_preview_button: (localStorage.getItem('wb.flag.preview')||'A')==='A' };
  return o;
})();