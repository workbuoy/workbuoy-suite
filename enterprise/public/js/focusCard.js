(function(){
  const list=document.getElementById('focus-suggestions');
  function why(evt){
    if(evt.type==='crm:renewal_due') return 'Fornyelse nærmer seg; høy churn‑risiko uten oppfølging.';
    if(evt.type==='crm:deal_stuck') return 'Sak har stått stille; anbefaler neste steg.';
    if(evt.type.startsWith('planner')) return 'Frist nær; forbered før møtet.';
    if(evt.type.startsWith('excel')) return 'Oppdaget mønster/feil i regneark.';
    return 'Relevant signal i kontekst.';
  }
  function rebuild(){
    const roleW = Roles.weights();
    const events=EventBus.last(80);
    const ranked = events.map(e=>({e,score:Policies.score(e, roleW)})).sort((a,b)=>b.score-a.score).slice(0,3);
    list.innerHTML='';
    ranked.forEach(({e})=>{
      const d=document.createElement('div'); d.className='msg';
      const title = e.title || (e.type.replace(':',' • '));
      const because = 'Jeg foreslår dette fordi: ' + why(e);
      const previewBtn = window.FLAGS.focus_preview_button ? '<button class="btn-outline" data-act="preview">Preview</button>' : '';
      d.innerHTML = `<div><strong>${title}</strong></div><div style="opacity:.8;font-size:12px">${because}</div><div style="margin-top:8px;display:flex;gap:8px"><button class="btn" data-act="apply">Bruk</button>${previewBtn}</div>`;
      d.querySelector('[data-act=apply]').onclick=()=>{
        EventBus.emit({type:'focus:apply', payload:e});
      };
      if(window.FLAGS.focus_preview_button){
        d.querySelector('[data-act=preview]').onclick=()=>{
          Overlay.show({content:{title:title,summary:e.summary||'',explanation:because}, suggestion:{actions:[{id:'apply',label:'Apply'},{id:'reject',label:'Reject'}]}, live_feed:false});
        };
      }
      list.appendChild(d);
    });
  }
  ['crm:renewal_due','crm:deal_stuck','planner:deadline_today','excel:formula_detected','excel:error_detected'].forEach(t=>EventBus.on(t,rebuild));
  EventBus.on('focus:refresh', rebuild);
  setInterval(rebuild, 3000);
})();