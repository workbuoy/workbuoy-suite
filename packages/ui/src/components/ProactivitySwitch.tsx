import type { KeyboardEvent } from "react";
import { useCallback, useMemo, useRef, useState } from "react";
import { motion } from "framer-motion";

export type Mode = "proactive" | "reactive";

export interface ProactivitySwitchProps {
  value?: Mode;
  defaultValue?: Mode;
  onChange?: (value: Mode) => void;
  disabled?: boolean;
  labels?: { proactive?: string; reactive?: string };
  size?: "sm" | "md";
  "aria-label"?: string;
  className?: string;
  id?: string;
}

const LABEL_DEFAULTS: Record<Mode, string> = {
  proactive: "Proactive",
  reactive: "Reactive",
};

const SIZE_STYLES: Record<NonNullable<ProactivitySwitchProps["size"]>, { container: string; button: string; text: string }> = {
  sm: {
    container: "p-0.5 text-sm",
    button: "px-2 py-1 text-xs",
    text: "text-xs",
  },
  md: {
    container: "p-1",
    button: "px-3 py-1.5 text-sm",
    text: "text-sm",
  },
};

const HIGHLIGHT_TRANSITION = {
  type: "spring" as const,
  stiffness: 500,
  damping: 35,
  mass: 0.8,
};

export function ProactivitySwitch({
  value,
  defaultValue = "reactive",
  onChange,
  disabled = false,
  labels,
  size = "md",
  className,
  id,
  "aria-label": ariaLabel,
}: ProactivitySwitchProps) {
  const isControlled = value !== undefined;
  const [internalValue, setInternalValue] = useState<Mode>(defaultValue);
  const resolvedValue = isControlled ? (value as Mode) : internalValue;

  const reactiveRef = useRef<HTMLButtonElement>(null);
  const proactiveRef = useRef<HTMLButtonElement>(null);

  const mergedLabels = useMemo(
    () => ({
      ...LABEL_DEFAULTS,
      ...labels,
    }),
    [labels],
  );

  const currentSize = SIZE_STYLES[size];

  const containerClassName = useMemo(() => {
    const base = "inline-flex items-center rounded-full border border-border bg-muted/50";
    const disabledClasses = disabled ? " opacity-60 cursor-not-allowed" : "";
    return [base, currentSize.container, "gap-1", "transition", className, disabledClasses]
      .filter(Boolean)
      .join(" ");
  }, [className, currentSize.container, disabled]);

  const selectMode = useCallback(
    (next: Mode, options?: { focus?: boolean }) => {
      if (disabled || resolvedValue === next) {
        if (options?.focus) {
          const ref = next === "reactive" ? reactiveRef.current : proactiveRef.current;
          ref?.focus();
        }
        return;
      }

      if (!isControlled) {
        setInternalValue(next);
      }

      onChange?.(next);

      if (options?.focus) {
        const ref = next === "reactive" ? reactiveRef.current : proactiveRef.current;
        ref?.focus();
      }
    },
    [disabled, isControlled, onChange, resolvedValue],
  );

  const handleKeyDown = useCallback(
    (event: KeyboardEvent<HTMLDivElement>) => {
      if (disabled) {
        return;
      }

      if (event.key === "ArrowRight") {
        event.preventDefault();
        selectMode("proactive", { focus: true });
      } else if (event.key === "ArrowLeft") {
        event.preventDefault();
        selectMode("reactive", { focus: true });
      }
    },
    [disabled, selectMode],
  );

  const handleSegmentKeyDown = useCallback(
    (segment: Mode) => (event: KeyboardEvent<HTMLButtonElement>) => {
      if (disabled) {
        event.preventDefault();
        return;
      }

      if (event.key === "Enter" || event.key === " ") {
        event.preventDefault();
        const next = segment === "reactive" ? "proactive" : "reactive";
        selectMode(next, { focus: true });
      }
    },
    [disabled, selectMode],
  );

  const renderButton = (segment: Mode) => {
    const isSelected = resolvedValue === segment;
    const ref = segment === "reactive" ? reactiveRef : proactiveRef;
    const label = mergedLabels[segment];
    const buttonClasses = [
      "relative flex-1 rounded-full font-medium transition",
      "focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background",
      "overflow-hidden text-center disabled:pointer-events-none",
      currentSize.button,
      isSelected ? "text-primary-foreground" : "text-muted-foreground",
    ]
      .filter(Boolean)
      .join(" ");

    return (
      <button
        key={segment}
        ref={ref}
        type="button"
        className={buttonClasses}
        aria-pressed={isSelected}
        aria-label={label}
        data-selected={isSelected ? true : undefined}
        disabled={disabled}
        tabIndex={isSelected ? 0 : -1}
        onClick={() => selectMode(segment, { focus: true })}
        onKeyDown={handleSegmentKeyDown(segment)}
      >
        {isSelected && (
          <motion.span
            layoutId="proactivity-switch-highlight"
            className="absolute inset-0 z-0 rounded-full bg-primary shadow"
            transition={HIGHLIGHT_TRANSITION}
            aria-hidden
          />
        )}
        <span className={`relative z-10 ${currentSize.text}`}>{label}</span>
      </button>
    );
  };

  const groupLabel = ariaLabel ?? "Proactivity mode";

  return (
    <div
      role="group"
      aria-label={groupLabel}
      className={containerClassName}
      id={id}
      data-disabled={disabled ? "true" : undefined}
      onKeyDown={handleKeyDown}
    >
      {renderButton("reactive")}
      {renderButton("proactive")}
    </div>
  );
}

export default ProactivitySwitch;
