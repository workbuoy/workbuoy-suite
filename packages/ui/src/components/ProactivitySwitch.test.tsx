import { act, render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";

import { ProactivitySwitch } from "./ProactivitySwitch.js";

describe("ProactivitySwitch", () => {
  it("renders with default reactive value", () => {
    render(<ProactivitySwitch />);

    const reactiveButton = screen.getByRole("button", { name: /reactive/i });
    const proactiveButton = screen.getByRole("button", { name: /proactive/i });

    expect(reactiveButton).toHaveAttribute("aria-pressed", "true");
    expect(proactiveButton).toHaveAttribute("aria-pressed", "false");
  });

  it("toggles to proactive on click and calls onChange", async () => {
    const user = userEvent.setup();
    const handleChange = vi.fn();

    render(<ProactivitySwitch onChange={handleChange} />);

    const proactiveButton = screen.getByRole("button", { name: /proactive/i });

    await act(async () => {
      await user.click(proactiveButton);
    });

    expect(proactiveButton).toHaveAttribute("aria-pressed", "true");
    expect(handleChange).toHaveBeenCalledWith("proactive");
  });

  it("handles keyboard arrow navigation", async () => {
    const user = userEvent.setup();

    render(<ProactivitySwitch />);

    await act(async () => {
      await user.tab();
    });
    const reactiveButton = screen.getByRole("button", { name: /reactive/i });
    expect(reactiveButton).toHaveFocus();

    await act(async () => {
      await user.keyboard("{ArrowRight}");
    });
    const proactiveButton = screen.getByRole("button", { name: /proactive/i });
    expect(proactiveButton).toHaveAttribute("aria-pressed", "true");
    expect(proactiveButton).toHaveFocus();

    await act(async () => {
      await user.keyboard("{ArrowLeft}");
    });
    expect(reactiveButton).toHaveAttribute("aria-pressed", "true");
    expect(reactiveButton).toHaveFocus();
  });

  it("toggles with space and enter", async () => {
    const user = userEvent.setup();

    render(<ProactivitySwitch />);

    await act(async () => {
      await user.tab();
    });
    const reactiveButton = screen.getByRole("button", { name: /reactive/i });

    await act(async () => {
      await user.keyboard(" ");
    });
    const proactiveButton = screen.getByRole("button", { name: /proactive/i });
    expect(proactiveButton).toHaveAttribute("aria-pressed", "true");

    await act(async () => {
      await user.keyboard("{Enter}");
    });
    expect(reactiveButton).toHaveAttribute("aria-pressed", "true");
  });

  it("notifies change in controlled mode without updating selection", async () => {
    const user = userEvent.setup();
    const handleChange = vi.fn();

    render(<ProactivitySwitch value="reactive" onChange={handleChange} />);

    const proactiveButton = screen.getByRole("button", { name: /proactive/i });

    await act(async () => {
      await user.click(proactiveButton);
    });

    expect(handleChange).toHaveBeenCalledWith("proactive");
    expect(proactiveButton).toHaveAttribute("aria-pressed", "false");
  });

  it("sets appropriate accessibility attributes", () => {
    render(<ProactivitySwitch />);

    const group = screen.getByRole("group", { name: /proactivity mode/i });
    const reactiveButton = screen.getByRole("button", { name: /reactive/i });
    const proactiveButton = screen.getByRole("button", { name: /proactive/i });

    expect(group).toBeInTheDocument();
    expect(reactiveButton).toHaveAttribute("aria-pressed", "true");
    expect(proactiveButton).toHaveAttribute("aria-pressed", "false");
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<ProactivitySwitch />);

    expect(asFragment()).toMatchSnapshot();
  });
});
