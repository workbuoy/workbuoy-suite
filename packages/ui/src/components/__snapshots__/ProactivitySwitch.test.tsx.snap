// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProactivitySwitch > matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-label="Proactivity mode"
    class="inline-flex items-center rounded-full border border-border bg-muted/50 p-1 gap-1 transition"
    role="group"
  >
    <button
      aria-label="Reactive"
      aria-pressed="true"
      class="relative flex-1 rounded-full font-medium transition focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background overflow-hidden text-center disabled:pointer-events-none px-3 py-1.5 text-sm text-primary-foreground"
      data-selected="true"
      tabindex="0"
      type="button"
    >
      <span
        aria-hidden="true"
        class="absolute inset-0 z-0 rounded-full bg-primary shadow"
      />
      <span
        class="relative z-10 text-sm"
      >
        Reactive
      </span>
    </button>
    <button
      aria-label="Proactive"
      aria-pressed="false"
      class="relative flex-1 rounded-full font-medium transition focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background overflow-hidden text-center disabled:pointer-events-none px-3 py-1.5 text-sm text-muted-foreground"
      tabindex="-1"
      type="button"
    >
      <span
        class="relative z-10 text-sm"
      >
        Proactive
      </span>
    </button>
  </div>
</DocumentFragment>
`;
